avr-hal-generic/src/port.rs

```diff
@@ -523,169 +521,176 @@ macro_rules! impl_port_traditional {
             }
-        )+
+        }
+
+        $crate::paste::paste! {
+            $($(
+                pub struct [<P $name $pin>] {
+                    _private: ()
+                }
+
+                impl $crate::port::PinOps for [<P $name $pin>] {
+                    type Dynamic = Dynamic;
+
+                    #[inline]
+                    fn into_dynamic(self) -> Self::Dynamic {
+                        Dynamic::new(DynamicPort::[<PORT $name>], $pin)
+                    }
+
+                    #[inline]
+                    unsafe fn out_set(&mut self) {
+                        (*<$port>::ptr()).[<port $name:lower>].modify(|r, w| {
+                            w.bits(r.bits() | (1 << $pin))
+                        })
+                    }
+
+                    #[inline]
+                    unsafe fn out_clear(&mut self) {
+                        (*<$port>::ptr()).[<port $name:lower>].modify(|r, w| {
+                            w.bits(r.bits() & !(1 << $pin))
+                        })
+                    }
+
+                    #[inline]
+                    unsafe fn out_toggle(&mut self) {
+                        (*<$port>::ptr()).[<pin $name:lower>].write(|w| w.bits(1 << $pin))
+                    }
+
+                    #[inline]
+                    unsafe fn out_get(&self) -> bool {
+                        (*<$port>::ptr()).[<port $name:lower>].read().bits() & (1 << $pin) != 0
+                    }
+
+                    #[inline]
+                    unsafe fn in_get(&self) -> bool {
+                        (*<$port>::ptr()).[<pin $name:lower>].read().bits() & (1 << $pin) != 0
+                    }
+
+                    #[inline]
+                    unsafe fn make_output(&mut self) {
+                        (*<$port>::ptr()).[<ddr $name:lower>].modify(|r, w| {
+                            w.bits(r.bits() | (1 << $pin))
+                        })
+                    }
+
+                    #[inline]
+                    unsafe fn make_input(&mut self, pull_up: bool) {
+                        (*<$port>::ptr()).[<ddr $name:lower>].modify(|r, w| {
+                            w.bits(r.bits() & !(1 << $pin))
+                        });
+                        if pull_up {
+                            self.out_set()
+                        } else {
+                            self.out_clear()
+                        }
+                    }
+                }
+            )+)+
+        }
     };
 }
 
 #[macro_export]
 macro_rules! renamed_pins {
     (
-        type Pin = $PinType:ident;
-
         $(#[$pins_attr:meta])*
-        pub struct Pins from $McuPins:ty {
-            $($(#[$pin_attr:meta])* pub $pin:ident: $Pin:ty = $pin_orig:ident,)+
+        pub struct Pins {
+            $($(#[$pin_attr:meta])* pub $pin_name:ident: $pin_type:ty = $pin_orig:ident,)+
+        }
+
+        impl Pins {
+            type Pin = $pin_wrapper:ident;
+            type McuPins = $mcu_pins:ty;
         }
```

Comment:
Can you comment on the motivation behind this change?  I don't have strong opinions either direction, but would like to hear why you prefer this macro syntax.

Reply:
This is indeed a more stylistic change. Now that I think about it, I could probably have left this bit untouched, just changing the transcriber. But in general, when writing macros I prefer to keep syntax as close to the base language as possible, and the `Pins from $McuPins` part both introduces a "new" keyword (`from`) and looks more like a syntax for inheritance (in the OOP sense). The type that the macro generates doesn't really inherit the other, in the sense of "class A is a B and therefore can do what B can and be where B would", it's more like they're alternative interfaces. I thought making it seem as an associated type would be more fitting, because 1) it still shows the relationship between the types; 2) it doesn't add any syntax (though inherent associated types is an unstable grammar, I believe, so maybe we could improve); 3) it looks more native to Rust, since it doesn't have proper inheritance.

All that being said, it's mostly superfluous, so if you'd rather make the diff a bit smaller I can change it back.

Reply:
Fair, thanks a lot for the detailed elaboration :)  Sounds good to me!

> All that being said, it's mostly superfluous, so if you'd rather make the diff a bit smaller I can change it back.

I think for the future, it would be a good idea to at least put each of the changes into its own commit as they don't logically relate to one another.  But let's not bother with such perfectionism here, I'll take your changes as they stand right now.

Reply:
> I think for the future, it would be a good idea to at least put each of the changes into its own commit as they don't logically relate to one another.

I can apply this suggestion, just have a question of preference before: Do you value the ability to compile each commit separately, or having it working by the end of the pull request is enough? There are some instances where using a more micro approach will require introducing changes that would not need to be there otherwise, to maintain self-containment. I'm fine with either, just would rather ask to make sure.